typescript/cli/src/rebalancer/monitor/Monitor.ts

```diff
@@ -0,0 +1,148 @@
+import EventEmitter from 'events';
+
+import type { Token, WarpCore } from '@hyperlane-xyz/sdk';
+import { sleep } from '@hyperlane-xyz/utils';
+
+import {
+  type IMonitor,
+  type MonitorEvent,
+  MonitorEventType,
+  MonitorPollingError,
+  MonitorStartError,
+} from '../interfaces/IMonitor.js';
+import { monitorLogger } from '../utils/logger.js';
+
+/**
+ * Simple monitor implementation that polls warp route collateral balances and emits them as MonitorEvent.
+ */
+export class Monitor implements IMonitor {
+  private readonly emitter = new EventEmitter();
+  private isMonitorRunning = false;
+
+  /**
+   * @param checkFrequency - The frequency to poll balances in ms.
+   */
+  constructor(
+    private readonly checkFrequency: number,
+    private readonly warpCore: WarpCore,
+  ) {}
+
+  // overloads from IMonitor
+  on(
+    eventName: MonitorEventType.TokenInfo,
+    fn: (event: MonitorEvent) => void,
+  ): this;
+  on(eventName: MonitorEventType.Error, fn: (event: Error) => void): this;
+  on(eventName: MonitorEventType.Start, fn: () => void): this;
+  on(eventName: string, fn: (...args: any[]) => void): this {
+    this.emitter.on(eventName, fn);
+    return this;
+  }
+
+  async start() {
+    if (this.isMonitorRunning) {
+      // Cannot start the same monitor multiple times
+      this.emitter.emit(
+        MonitorEventType.Error,
+        new MonitorStartError('Monitor already running'),
+      );
+      return;
+    }
+
+    try {
+      this.isMonitorRunning = true;
+      monitorLogger.debug(
+        { checkFrequency: this.checkFrequency },
+        'Monitor started',
+      );
+      this.emitter.emit(MonitorEventType.Start);
+
+      while (this.isMonitorRunning) {
+        try {
+          monitorLogger.debug('Polling cycle started');
+          const event: MonitorEvent = {
+            tokensInfo: [],
+          };
+
+          for (const token of this.warpCore.tokens) {
+            monitorLogger.debug(
+              {
+                chain: token.chainName,
+                tokenSymbol: token.symbol,
+                tokenAddress: token.addressOrDenom,
+              },
+              'Checking token',
+            );
+            const bridgedSupply = await this.getTokenBridgedSupply(token);
```

Comment:
This may be a subtle bug. The bridged supply could be undefined at this point. We only log a warning about this in this file. But later on in `getRawBalances()` we throw an exception if an undefined value is encountered.